### PR TITLE
Wownero integration #2

### PR DESCRIPTION
**Pull request by sanderfoobar**
Continuation of https://github.com/cake-tech/cake_wallet/pull/362

Wownero support. Uses Polyseed this time (we are phasing out `wownero-seed`).

I tested: sending/receiving/history on both Android (arm64) and Android emulator (x86_64).

Screenshot from 2024-01-03 13-59-59

## Notes:

There are still some to-do's:

### 1. Fiat <-> WOW calculations

Probably requires https://github.com/cake-tech/cake_wallet/issues/361

### 2. Accounts runtime error

`Dashboard -> Receive -> Accounts` gives an error. Fix would be to leave out the accounts feature in its entirety just like last PR - as we do not use the account feature in Wownero. Would be nice if someone could exclude this feature when `coin type == wownero`.

### 3. Exchange button

The exchange button at the dashboard for WOW should (probably) not be active.

### 4. iOS/macos untested

Untested - getting the iOS emulator to run is a bit difficult with the certificates and all.

### 5. pubspec

I updated `watcher` to `1.1.0`.

### 6. Docker fixes

The following had some issues:

```text
cd scripts/docker
docker build --tag=build_monero_deps --progress plain --no-cache .
docker compose up
```

I fixed various things in `scripts/docker`:

- Changed file permission mode
- `build_haven.sh` interpreter changed to `/bin/bash`
- Added build requirement `byacc` to `apt get` in `Dockerfile`, it is needed for libunbound to compile correctly
- Changed carriage return to Unix style (sorry for the diffs)

### 7. Wownero branch

Changes to Wownero to facilitate Cake using wow's wallet2 API went into a branch called [cake-fixes](https://git.wownero.com/wownero/wownero/commits/branch/cake-fixes). The build script `build_wownero.sh` is pinned to a commit hash.

### 8. Polyseed

Waiting on:

- https://github.com/tevador/polyseed/pull/10
- https://github.com/cake-tech/polyseed_dart/pull/1

Current situation is that create wallet generates a Polyseed with the Monero coin type.

### CODE REVIEW

**Review by nahuhh — COMMENTED** [review]
Just spotted a couple places that were 12 instead of 11

File: cw_core/lib/node.dart [hunk 1]
```diff
@@ -134,6 +136,7 @@ class Node extends HiveObject with Keyable {
     try {
       switch (type) {
         case WalletType.monero:
+        case WalletType.wownero:
           return requestMoneroNode();
         case WalletType.bitcoin:
           return requestElectrumServer();
```
Comment (nahuhh) [inline cw_core/lib/node.dart:142]:
```suggestion
         case WalletType.monero:
         case WalletType.wownero:
         case WalletType.haven:
           return requestMoneroNode();
         case WalletType.bitcoin:
           return requestElectrumServer();
```

perhaps ^ + del L145-L146 

(missing a space / tab is misaligned for the wownero case line)

**Comment by nahuhh** [comment]
>3. Exchange button
>
>The exchange button at the dashboard for WOW should (probably) not be active.

trocador supports wownero iirc. So i think it can stay (would need to add wownero to the list of swappable coins)

**Comment by nahuhh** [comment]
@OmarHatem28  please look into

> ### 1. Fiat <-> WOW calculations
> 
> Probably requires https://github.com/cake-tech/cake_wallet/issues/361

>### 2. Accounts runtime error
> 
> `Dashboard -> Receive -> Accounts` gives an error. Fix would be to leave out the accounts feature in its entirety just like last PR - as we do not use the account feature in Wownero. Would be nice if someone could exclude this feature when `coin type == wownero`.
> 
> ### 3. Exchange button
> 
> ~~The exchange button at the dashboard for WOW should (probably) not be active.~~

Trocador supports wownero, so you should be able to add wow as a supported currency
> 
> ### 4. iOS/macos untested
> 
> Untested - getting the iOS emulator to run is a bit difficult with the certificates and all.

> ### 8. Polyseed
> 
> Waiting on:
> 
> - https://github.com/cake-tech/polyseed_dart/pull/1


Needs review / confirm everything ok:
> ### 5. pubspec
> 
> I updated `watcher` to `1.1.0`.
> 
> ### 6. Docker fixes
> 
> The following had some issues:
> 
> ```text
> cd scripts/docker
> docker build --tag=build_monero_deps --progress plain --no-cache .
> docker compose up
> ```
> 
> I fixed various things in `scripts/docker`:
> 
> - Changed file permission mode
> - `build_haven.sh` interpreter changed to `/bin/bash`
> - Added build requirement `byacc` to `apt get` in `Dockerfile`, it is needed for libunbound to compile correctly
> - Changed carriage return to Unix style (sorry for the diffs)
> 
> ### 7. Wownero branch
> 
> Changes to Wownero to facilitate Cake using wow's wallet2 API went into a branch called [cake-fixes](https://git.wownero.com/wownero/wownero/commits/branch/cake-fixes). The build script `build_wownero.sh` is pinned to a commit hash.
> 



> Current situation is that create wallet generates a Polyseed with the Monero coin type.

need to change to wownero after polyseed_dart pr merge

File: assets/wownero_node_list.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  is_default: true
+  useSSL: false
+-
+  uri: amsterdam.muchwow.lol:34568
+  is_default: false
+  useSSL: false
+-
+  uri: singapore.muchwow.lol:34568
+  is_default: false
+  useSSL: false
```
Comment (nahuhh) [inline assets/wownero_node_list.yml:12]:
```suggestion
  uri: node3.monerodevs.org:34568
  is_default: true
  useSSL: false
-
  uri: node2.monerodevs.org:34568
  is_default: false
  useSSL: false
-
  uri: node.monerodevs.org:34568
  is_default: false
  useSSL: false
```

**Review by nahuhh — CHANGES_REQUESTED** [review]

**Comment by OmarHatem28** [comment]
Hi @nahuhh thanks for helping in reviewing this PR

I am currently in the review process but I might submit my current comments so far until I complete the review.

I will review the Polyseed PR afterwards but @sanderfoobar can't we make both seed types (normal, polyseed) supported?

Also, like @nahuhh suggested, can you please remove the Accounts feature and all its related files, and update the PR with the latest main to resolve the conflicts

**Review by OmarHatem28 — CHANGES_REQUESTED** [review]
No need for `android/app/src/main/java/com/cakewallet/wownero/` since there won't be a separate app for wownero and it would be included in Cake Wallet

File: lib/src/screens/seed/pre_seed_page.dart [hunk 3]
```diff
@@ -86,6 +86,8 @@ class PreSeedPage extends BasePage {
         if (moneroSeedType == SeedType.polyseed)
           return 16;
         return 25;
+      case WalletType.wownero:
+        return 16;  // always polyseed
```
[the remark this answers is not in the document]
Comment (nahuhh) [reply lib/src/screens/seed/pre_seed_page.dart:90]: currently using monero's polyseed (so likely creates invalid seeds)

**Comment by OmarHatem28** [comment]
Hi @sanderfoobar We're excited about the changes you've proposed and adding Wownero, and we would love to see this PR integrated into the Cake.
If you have any questions or need further assistance with your pull request, please don't hesitate to let me know.

**Comment by nahuhh** [comment]
@OmarHatem28 to proceed and properly test, from cake side, we need https://github.com/cake-tech/polyseed_dart/pull/1

**Comment by OmarHatem28** [comment]
@nahuhh merged

File: cw_core/lib/crypto_currency.dart [hunk 4]
```diff
@@ -217,7 +218,7 @@ class CryptoCurrency extends EnumerableItem<int> with Serializable<int> implemen
   static const kaspa = CryptoCurrency(title: 'KAS', fullName: 'Kaspa', raw: 89, name: 'kas', iconPath: 'assets/images/kaspa_icon.png', decimals: 8);
   static const digibyte = CryptoCurrency(title: 'DGB', fullName: 'DigiByte', raw: 90, name: 'dgb', iconPath: 'assets/images/digibyte.png', decimals: 8);
   static const usdtSol = CryptoCurrency(title: 'USDT', tag: 'SOL', fullName: 'USDT Tether', raw: 91, name: 'usdtsol', iconPath: 'assets/images/usdt_icon.png', decimals: 6);
-
+  static const wow = CryptoCurrency(title: 'WOW', fullName: 'Wownero', raw: 89, name: 'wow', iconPath: 'assets/images/wownero_icon.png', decimals: 11);
```
Comment (nahuhh) [inline cw_core/lib/crypto_currency.dart:221]:
```suggestion
  static const wow = CryptoCurrency(title: 'WOW', fullName: 'Wownero', raw: 92, name: 'wow', iconPath: 'assets/images/wownero_icon.png', decimals: 11);
```

File: cw_wownero/ios/Classes/wownero_api.cpp [hunk 5]
```diff
@@ -0,0 +1,1034 @@
...
+}
+
+void account_set_label_row(uint32_t account_index, char *label)
+{
+    m_account->setLabel(account_index, label);
+}
+
+void account_refresh()
+{
+    m_account->refresh();
```
Comment (nahuhh) [inline cw_wownero/ios/Classes/wownero_api.cpp:849]: @OmarHatem28 
idk where to remove accounts. Can you guys handle

1. Removing accounts
2. Adding to fiat api
3. Adding to exchange

**Review by nahuhh — CHANGES_REQUESTED** [review]

**Review by nahuhh — APPROVED** [review]
soon, no more tm

**Comment by nahuhh** [comment]
need from cake for release:
- [ ] ios test
- [ ] fiat api 
- [ ] exchange integration
- [ ] wownero easter eggs (jk.)

**Comment by nahuhh** [comment]
Pinging @OmarHatem28

**Comment by MrCyjaneK** [comment]
Hey! There is a little detail that I've encountered when a combination of wownero and monero is used together in one process as a shared library.

The issue is as follows
- once you open wownero wallet it works great
- when you try to use monero wallet it uses previously loaded wownero symbols
- so when you opened one of the coins, another one may reuse it's symbols.

I've encountered this issue on linux while working on another wallet that uses cw libraries, I'm not sure if this is the case with cake in this PR (I'll check it once I have some more time), but this is something that may cause issue.

**Comment by OmarHatem28** [comment]
@MrCyjaneK thanks for sharing this issue, we will check if it's indeed the case in #1345, alongside the other fixes, feel free to try the APK in there and share your insights on the PR if you find anything missing or should be fixed or needs enhancements.
we will try to give more attention to finalizing Wownero soon if everything goes smooth

**Comment by MrCyjaneK** [comment]
@sanderfoobar up to my current knowledge it is.

**Comment by OmarHatem28** [comment]
Hey, thanks for all the contributions so far, Wownero is now in Beta with this PR #1485 
please feel free to try it out, and send us your feedback.
will be closing this in favor of #1485

**Pull request closed without merging**